File: cmd/spacecraft/map.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

func mapCmd(args []string, spaceDir, _ string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map <new|add|rm|ls|show|next|use|current|archive> [...]")
		return 1
	}

	roadmapsDir := filepath.Join(spaceDir, "roadmaps")
	os.MkdirAll(roadmapsDir, 0755)

	switch args[0] {
	case "new":
		return mapNew(args[1:], roadmapsDir)
	case "add":
		return mapAdd(args[1:], roadmapsDir)
	case "rm":
		return mapRemove(args[1:], roadmapsDir)
	case "ls":
		return mapList(roadmapsDir)
	case "show":
		return mapShow(args[1:], roadmapsDir)
	case "next":
		return mapNext(args[1:], roadmapsDir)
	case "use":
		return mapUse(args[1:], spaceDir, roadmapsDir)
	case "current":
		return mapCurrent(spaceDir)
	case "archive":
		return mapArchive(args[1:], roadmapsDir, spaceDir)
	default:
		fmt.Fprintf(os.Stderr, "spacecraft map: unknown subcommand %q\n", args[0])
		return 1
	}
}

func mapNew(args []string, dir string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map new <title> [--desc <text>]")
		return 1
	}
	title := args[0]
	desc := ""
	for i := 1; i < len(args)-1; i++ {
		if args[i] == "--desc" {
			desc = args[i+1]
		}
	}

	id := strings.ToLower(strings.ReplaceAll(title, " ", "-"))
	id = strings.Map(func(r rune) rune {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '-' {
			return r
		}
		return -1
	}, id)

	now := time.Now().UTC().Format(time.RFC3339)
	r := map[string]interface{}{
		"id":          id,
		"title":       title,
		"description": desc,
		"missions":    []interface{}{},
		"issues":      []interface{}{},
		"createdAt":   now,
		"updatedAt":   now,
	}

	return saveRoadmap(dir, id, r)
}

func mapAdd(args []string, dir string) int {
	if len(args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map add <roadmap-id> <mission-id> [--desc <text>]")
		return 1
	}
	rid, mid := args[0], args[1]
	desc := mid
	for i := 2; i < len(args)-1; i++ {
		if args[i] == "--desc" {
			desc = args[i+1]
		}
	}

	r, err := loadRoadmap(dir, rid)
	if err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}

	missions, _ := r["missions"].([]interface{})
	r["missions"] = append(missions, map[string]interface{}{
		"id":          mid,
		"description": desc,
	})
	r["updatedAt"] = time.Now().UTC().Format(time.RFC3339)

	return saveRoadmap(dir, rid, r)
}

func mapRemove(args []string, dir string) int {
	if len(args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map rm <roadmap-id> <mission-id>")
		return 1
	}
	rid, mid := args[0], args[1]

	r, err := loadRoadmap(dir, rid)
	if err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}

	missions, _ := r["missions"].([]interface{})
	var filtered []interface{}
	for _, m := range missions {
		entry, _ := m.(map[string]interface{})
		if entry["id"] != mid {
			filtered = append(filtered, m)
		}
	}
	r["missions"] = filtered
	r["updatedAt"] = time.Now().UTC().Format(time.RFC3339)

	return saveRoadmap(dir, rid, r)
}

func mapList(dir string) int {
	entries, _ := os.ReadDir(dir)
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		id := strings.TrimSuffix(e.Name(), ".json")
		r, err := loadRoadmap(dir, id)
		if err != nil {
			continue
		}
		missions, _ := r["missions"].([]interface{})
		issues, _ := r["issues"].([]interface{})
		fmt.Printf("%-30s %s (%d missions, %d issues)\n",
			id, r["title"], len(missions), len(issues))
	}
	return 0
}

func mapShow(args []string, dir string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map show <roadmap-id>")
		return 1
	}
	r, err := loadRoadmap(dir, args[0])
	if err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}
	out, _ := json.MarshalIndent(r, "", "  ")
	fmt.Println(string(out))
	return 0
}

func mapNext(args []string, dir string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map next <roadmap-id>")
		return 1
	}
	r, err := loadRoadmap(dir, args[0])
	if err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}
	spaceDir := filepath.Join(dir, "..")
	if id, desc, state, ok := nextIncompleteOnRoadmap(spaceDir, r); ok {
		fmt.Printf("%s: %s (state=%s)\n", id, desc, state)
		return 0
	}
	fmt.Println("All missions complete.")
	return 0
}

// nextIncompleteOnRoadmap returns the first incomplete (non-ready, non-shipped,
// non-archived) mission on a roadmap.
func nextIncompleteOnRoadmap(spaceDir string, r map[string]interface{}) (id, desc, state string, ok bool) {
	missions, _ := r["missions"].([]interface{})
	for _, m := range missions {
		entry, _ := m.(map[string]interface{})
		mid, _ := entry["id"].(string)
		d, _ := entry["description"].(string)
		if mid == "" {
			continue
		}
		if _, err := os.Stat(filepath.Join(spaceDir, "archive", mid)); err == nil {
			continue
		}
		st := missionIncompleteState(filepath.Join(spaceDir, "missions", mid))
		if st == "" {
			continue
		}
		return mid, d, st, true
	}
	return "", "", "", false
}

func roadmapContainsMission(r map[string]interface{}, missionID string) bool {
	missions, _ := r["missions"].([]interface{})
	for _, m := range missions {
		entry, _ := m.(map[string]interface{})
		if id, _ := entry["id"].(string); id == missionID {
			return true
		}
	}
	return false
}

// findRoadmapForArchivedMission prefers current-roadmap when it lists the
// archived mission; otherwise scans roadmaps for one that contains it.
func findRoadmapForArchivedMission(spaceDir, archivedID string) (rid string, r map[string]interface{}, ok bool) {
	roadmapsDir := filepath.Join(spaceDir, "roadmaps")
	if data, err := os.ReadFile(filepath.Join(spaceDir, "current-roadmap")); err == nil {
		cur := strings.TrimSpace(string(data))
		if cur != "" {
			if rm, err := loadRoadmap(roadmapsDir, cur); err == nil && roadmapContainsMission(rm, archivedID) {
				return cur, rm, true
			}
		}
	}
	entries, err := os.ReadDir(roadmapsDir)
	if err != nil {
		return "", nil, false
	}
	var ids []string
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		ids = append(ids, strings.TrimSuffix(e.Name(), ".json"))
	}
	sort.Strings(ids)
	for _, id := range ids {
		rm, err := loadRoadmap(roadmapsDir, id)
		if err != nil {
			continue
		}
		if roadmapContainsMission(rm, archivedID) {
			return id, rm, true
		}
	}
	return "", nil, false
}

// suggestNextAfterArchive clears stale current, advances to the next incomplete
// roadmap mission when one exists, and prints a human hint.
func suggestNextAfterArchive(spaceDir, archivedID string) {
	if readCurrent(spaceDir) == archivedID {
		clearCurrent(spaceDir)
	}
	rid, r, found := findRoadmapForArchivedMission(spaceDir, archivedID)
	if !found {
		return
	}
	nextID, desc, state, ok := nextIncompleteOnRoadmap(spaceDir, r)
	if !ok {
		return
	}
	_ = writeCurrent(spaceDir, nextID)
	fmt.Printf("Next mission on roadmap %s: %s: %s (state=%s)\n", rid, nextID, desc, state)
	fmt.Printf("Suggested: new session → /sc-discuss %s (then /sc-run)\n", nextID)
}

// missionIncompleteState returns the display state for an incomplete mission, or "" if complete.
// Missing mission dir / mission.json → "pending". ready/shipped → complete ("").
func missionIncompleteState(missionPath string) string {
	data, err := os.ReadFile(filepath.Join(missionPath, "mission.json"))
	if err != nil {
		return "pending"
	}
	var mj map[string]interface{}
	if err := json.Unmarshal(data, &mj); err != nil {
		return "pending"
	}
	state, _ := mj["state"].(string)
	if state == "" {
		return "pending"
	}
	if state == "ready" || state == "shipped" {
		return ""
	}
	return state
}

func mapUse(args []string, spaceDir, roadmapsDir string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map use <roadmap-id>")
		return 1
	}
	id := args[0]
	if _, err := loadRoadmap(roadmapsDir, id); err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}
	if err := os.MkdirAll(spaceDir, 0755); err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}
	if err := os.WriteFile(filepath.Join(spaceDir, "current-roadmap"), []byte(id+"\n"), 0644); err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}
	fmt.Printf("Selected roadmap %s\n", id)
	return 0
}

func mapCurrent(spaceDir string) int {
	data, err := os.ReadFile(filepath.Join(spaceDir, "current-roadmap"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map: no current roadmap")
		return 1
	}
	id := strings.TrimSpace(string(data))
	if id == "" {
		fmt.Fprintln(os.Stderr, "spacecraft map: no current roadmap")
		return 1
	}
	fmt.Println(id)
	return 0
}

func mapArchive(args []string, dir string, spaceDir string) int {
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "Usage: spacecraft map archive <roadmap-id>")
		return 1
	}
	id := args[0]
	archiveDir := filepath.Join(spaceDir, "archive", "roadmaps")
	os.MkdirAll(archiveDir, 0755)

	src := filepath.Join(dir, id+".json")
	dst := filepath.Join(archiveDir, id+".json")
	if err := os.Rename(src, dst); err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}
	fmt.Printf("Archived roadmap: %s\n", id)
	return 0
}

func loadRoadmap(dir, id string) (map[string]interface{}, error) {
	data, err := os.ReadFile(filepath.Join(dir, id+".json"))
	if err != nil {
		return nil, fmt.Errorf("roadmap not found: %s", id)
	}
	var r map[string]interface{}
	if err := json.Unmarshal(data, &r); err != nil {
		return nil, fmt.Errorf("invalid roadmap: %v", err)
	}
	return r, nil
}

func saveRoadmap(dir, id string, r map[string]interface{}) int {
	data, _ := json.MarshalIndent(r, "", "  ")
	if err := os.WriteFile(filepath.Join(dir, id+".json"), append(data, '\n'), 0644); err != nil {
		fmt.Fprintln(os.Stderr, "spacecraft map:", err)
		return 1
	}
	fmt.Printf("Roadmap saved: %s\n", id)
	return 0
}

// Ensure sort imports work
var _ = sort.Strings

File: cli/lib/map.mjs
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  renameSync,
  unlinkSync,
  writeFileSync,
} from 'node:fs';
import path from 'node:path';
import { writeCurrent } from './mission.mjs';
import { currentFile, readCurrent } from './resolve.mjs';

function hasHelpFlag(args) {
  return args.some((a) => a === '--help' || a === '-h');
}

function roadmapPath(dir, id) {
  return path.join(dir, `${id}.json`);
}

function loadRoadmap(dir, id) {
  let data;
  try {
    data = readFileSync(roadmapPath(dir, id), 'utf8');
  } catch {
    throw new Error(`roadmap not found: ${id}`);
  }
  try {
    return JSON.parse(data);
  } catch (err) {
    throw new Error(`invalid roadmap: ${err.message}`);
  }
}

function saveRoadmap(dir, id, r) {
  try {
    writeFileSync(roadmapPath(dir, id), `${JSON.stringify(r, null, 2)}\n`);
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  console.log(`Roadmap saved: ${id}`);
  return 0;
}

function slugTitle(title) {
  return title
    .toLowerCase()
    .replaceAll(' ', '-')
    .replace(/[^a-z0-9-]/g, '');
}

function flagValue(args, name) {
  for (let i = 0; i < args.length - 1; i += 1) {
    if (args[i] === name) return args[i + 1];
  }
  return '';
}

function mapNew(args, dir) {
  if (hasHelpFlag(args)) {
    console.log('Usage: spacecraft map new <title> [--desc <text>]');
    return 0;
  }
  if (args.length < 1) {
    console.error('Usage: spacecraft map new <title> [--desc <text>]');
    return 1;
  }
  const title = args[0];
  const desc = flagValue(args.slice(1), '--desc');
  const id = slugTitle(title);
  const now = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
  const r = {
    id,
    title,
    description: desc,
    missions: [],
    issues: [],
    createdAt: now,
    updatedAt: now,
  };
  return saveRoadmap(dir, id, r);
}

function mapAdd(args, dir) {
  if (args.length < 2) {
    console.error('Usage: spacecraft map add <roadmap-id> <mission-id> [--desc <text>]');
    return 1;
  }
  const rid = args[0];
  const mid = args[1];
  const desc = flagValue(args.slice(2), '--desc') || mid;
  let r;
  try {
    r = loadRoadmap(dir, rid);
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  const missions = Array.isArray(r.missions) ? [...r.missions] : [];
  missions.push({ id: mid, description: desc });
  r.missions = missions;
  r.updatedAt = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
  return saveRoadmap(dir, rid, r);
}

function mapRemove(args, dir) {
  if (args.length < 2) {
    console.error('Usage: spacecraft map rm <roadmap-id> <mission-id>');
    return 1;
  }
  const rid = args[0];
  const mid = args[1];
  let r;
  try {
    r = loadRoadmap(dir, rid);
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  const missions = Array.isArray(r.missions) ? r.missions : [];
  r.missions = missions.filter((m) => m && m.id !== mid);
  r.updatedAt = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
  return saveRoadmap(dir, rid, r);
}

function mapList(dir) {
  let entries;
  try {
    entries = readdirSync(dir);
  } catch {
    return 0;
  }
  for (const name of entries) {
    if (!name.endsWith('.json')) continue;
    const id = name.slice(0, -'.json'.length);
    let r;
    try {
      r = loadRoadmap(dir, id);
    } catch {
      continue;
    }
    const missions = Array.isArray(r.missions) ? r.missions : [];
    const issues = Array.isArray(r.issues) ? r.issues : [];
    const title = typeof r.title === 'string' ? r.title : '';
    console.log(`${id.padEnd(30)} ${title} (${missions.length} missions, ${issues.length} issues)`);
  }
  return 0;
}

function mapShow(args, dir) {
  if (args.length < 1) {
    console.error('Usage: spacecraft map show <roadmap-id>');
    return 1;
  }
  let r;
  try {
    r = loadRoadmap(dir, args[0]);
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  console.log(JSON.stringify(r, null, 2));
  return 0;
}

/** Incomplete display state, or "" if ready/shipped (complete). Missing → pending. */
function missionIncompleteState(missionPath) {
  let data;
  try {
    data = readFileSync(path.join(missionPath, 'mission.json'), 'utf8');
  } catch {
    return 'pending';
  }
  let mj;
  try {
    mj = JSON.parse(data);
  } catch {
    return 'pending';
  }
  const state = typeof mj.state === 'string' ? mj.state : '';
  if (state === '') return 'pending';
  if (state === 'ready' || state === 'shipped') return '';
  return state;
}

export function nextIncompleteOnRoadmap(spaceDir, r) {
  const missions = Array.isArray(r.missions) ? r.missions : [];
  for (const entry of missions) {
    if (!entry || typeof entry.id !== 'string' || entry.id === '') continue;
    const mid = entry.id;
    const desc = typeof entry.description === 'string' ? entry.description : '';
    if (existsSync(path.join(spaceDir, 'archive', mid))) continue;
    const st = missionIncompleteState(path.join(spaceDir, 'missions', mid));
    if (st === '') continue;
    return { id: mid, desc, state: st };
  }
  return null;
}

function mapNext(args, dir) {
  if (args.length < 1) {
    console.error('Usage: spacecraft map next <roadmap-id>');
    return 1;
  }
  let r;
  try {
    r = loadRoadmap(dir, args[0]);
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  const spaceDir = path.join(dir, '..');
  const next = nextIncompleteOnRoadmap(spaceDir, r);
  if (next) {
    console.log(`${next.id}: ${next.desc} (state=${next.state})`);
    return 0;
  }
  console.log('All missions complete.');
  return 0;
}

function roadmapContainsMission(r, missionID) {
  const missions = Array.isArray(r.missions) ? r.missions : [];
  return missions.some((m) => m && m.id === missionID);
}

function findRoadmapForArchivedMission(spaceDir, archivedID) {
  const roadmapsDir = path.join(spaceDir, 'roadmaps');
  try {
    const cur = readFileSync(path.join(spaceDir, 'current-roadmap'), 'utf8').trim();
    if (cur) {
      try {
        const rm = loadRoadmap(roadmapsDir, cur);
        if (roadmapContainsMission(rm, archivedID)) {
          return { rid: cur, r: rm };
        }
      } catch {
        // fall through to scan
      }
    }
  } catch {
    // no current-roadmap
  }
  let entries;
  try {
    entries = readdirSync(roadmapsDir);
  } catch {
    return null;
  }
  const ids = entries
    .filter((name) => name.endsWith('.json'))
    .map((name) => name.slice(0, -'.json'.length))
    .sort();
  for (const id of ids) {
    try {
      const rm = loadRoadmap(roadmapsDir, id);
      if (roadmapContainsMission(rm, archivedID)) {
        return { rid: id, r: rm };
      }
    } catch {
      // skip
    }
  }
  return null;
}

function clearCurrent(spaceDir) {
  try {
    unlinkSync(currentFile(spaceDir));
  } catch {
    // absent is fine
  }
}

/** Clear stale current, advance to next incomplete roadmap mission, print hint. */
export function suggestNextAfterArchive(spaceDir, archivedID) {
  if (readCurrent(spaceDir) === archivedID) {
    clearCurrent(spaceDir);
  }
  const found = findRoadmapForArchivedMission(spaceDir, archivedID);
  if (!found) return;
  const next = nextIncompleteOnRoadmap(spaceDir, found.r);
  if (!next) return;
  try {
    writeCurrent(spaceDir, next.id);
  } catch {
    // still print hint
  }
  console.log(
    `Next mission on roadmap ${found.rid}: ${next.id}: ${next.desc} (state=${next.state})`,
  );
  console.log(`Suggested: new session → /sc-discuss ${next.id} (then /sc-run)`);
}

function mapUse(args, spaceDir, roadmapsDir) {
  if (args.length < 1) {
    console.error('Usage: spacecraft map use <roadmap-id>');
    return 1;
  }
  const id = args[0];
  try {
    loadRoadmap(roadmapsDir, id);
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  try {
    mkdirSync(spaceDir, { recursive: true });
    writeFileSync(path.join(spaceDir, 'current-roadmap'), `${id}\n`);
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  console.log(`Selected roadmap ${id}`);
  return 0;
}

function mapCurrent(spaceDir) {
  let data;
  try {
    data = readFileSync(path.join(spaceDir, 'current-roadmap'), 'utf8');
  } catch {
    console.error('spacecraft map: no current roadmap');
    return 1;
  }
  const id = data.trim();
  if (!id) {
    console.error('spacecraft map: no current roadmap');
    return 1;
  }
  console.log(id);
  return 0;
}

function mapArchive(args, dir, spaceDir) {
  if (args.length < 1) {
    console.error('Usage: spacecraft map archive <roadmap-id>');
    return 1;
  }
  const id = args[0];
  const archiveDir = path.join(spaceDir, 'archive', 'roadmaps');
  try {
    mkdirSync(archiveDir, { recursive: true });
    renameSync(path.join(dir, `${id}.json`), path.join(archiveDir, `${id}.json`));
  } catch (err) {
    console.error('spacecraft map:', err.message);
    return 1;
  }
  console.log(`Archived roadmap: ${id}`);
  return 0;
}

export function mapCmd(args, spaceDir) {
  if (args.length < 1) {
    console.error(
      'Usage: spacecraft map <new|add|rm|ls|show|next|use|current|archive> [...]',
    );
    return 1;
  }
  const roadmapsDir = path.join(spaceDir, 'roadmaps');
  mkdirSync(roadmapsDir, { recursive: true });
  const sub = args[0];
  const rest = args.slice(1);
  switch (sub) {
    case 'new':
      return mapNew(rest, roadmapsDir);
    case 'add':
      return mapAdd(rest, roadmapsDir);
    case 'rm':
    case 'remove':
      return mapRemove(rest, roadmapsDir);
    case 'ls':
    case 'list':
      return mapList(roadmapsDir);
    case 'show':
      return mapShow(rest, roadmapsDir);
    case 'next':
      return mapNext(rest, roadmapsDir);
    case 'use':
      return mapUse(rest, spaceDir, roadmapsDir);
    case 'current':
      return mapCurrent(spaceDir);
    case 'archive':
      return mapArchive(rest, roadmapsDir, spaceDir);
    default:
      console.error(`spacecraft map: unknown subcommand ${JSON.stringify(sub)}`);
      return 1;
  }
}
